### DISCUSSION TITLE
Absolute permeability

### DISCUSSION THREAD

**Discussion by HFenice** [PMEAL/OpenPNM#2304] [Q&A]
Hi, I have a problem with the absolute permeability.

Here is my code:

# Importing library

import math
import openpnm as op
import matplotlib.pyplot as plt
import numpy as np
import scipy.stats as spst
from scipy.optimize import curve_fit
import csv
import pandas
import random
import openpnm.models as mods
from openpnm.geometry import GenericGeometry
from openpnm.io import PerGeos

df = pandas.read_csv('Dati_crop.csv')

print(df)

diametro = df['Diametro'].values

frequenza = df['Frequenza'].values

diametro = diametro.reshape((len(diametro), 1))

diametro = diametro*(10**-6)

frequenza = frequenza.reshape((len(frequenza), 1))

plt.plot(diametro,frequenza)

plt.show

# Network topology

pts= np.random.rand(int(sum(frequenza)), 3)*[4.45*(10**-3),4.45*(10**-3),4.45*(10**-3)]

pn = op.network.Gabriel(points = pts,shape=[4.45*(10**-3),4.45*(10**-3),4.45*(10**-3)])

fig = op.topotools.plot_connections(pn)

# Checking network health

h = pn.check_network_health()
print(h)

# My Geometry object by modifing StickAndBall class

class MyGeo(GenericGeometry):
    def __init__(self, random_state = None, **kwargs):
        super().__init__(**kwargs)
        
        random.seed(random_state)

        self['pore.diameter'] = np.asarray(random.choices(diametro,            
                                                          frequenza, 
                                                          k = pn.Np)).ravel()

        self.add_model(propname='pore.area',
                       model=mods.geometry.pore_cross_sectional_area.sphere,
                       pore_diameter='pore.diameter') 

        self.add_model(propname='pore.volume',
                       model=mods.geometry.pore_volume.sphere, 
                       pore_diameter='pore.diameter')
        
        self['throat.diameter']=0.4*(10**-6)
     
        self.add_model(propname='throat.endpoints',
                       model=mods.geometry.throat_endpoints.spherical_pores,
                       pore_diameter='pore.diameter',
                       throat_diameter='throat.diameter')

        self.add_model(propname='throat.length',
                       model=mods.geometry.throat_length.spheres_and_cylinders, 
                       pore_diameter='pore.diameter',
                       throat_diameter='throat.diameter')

        self.add_model(propname='throat.surface_area',
                       model=mods.geometry.throat_surface_area.cylinder, 
                       throat_diameter='throat.diameter',
                       throat_length='throat.length')

        self.add_model(propname='throat.volume',
                       model=mods.geometry.throat_volume.cylinder, 
                       throat_diameter='throat.diameter',
                       throat_length='throat.length')

        self.add_model(propname='throat.area',
                       model=mods.geometry.throat_cross_sectional_area.cylinder, 
                       throat_diameter='throat.diameter')
        
        self.add_model(propname='throat.conduit_lengths',
                       model=mods.geometry.throat_length.conduit_lengths,  
                       throat_endpoints='throat.endpoints',
                       throat_length='throat.length')
        
        self.add_model(propname='throat.hydraulic_size_factors',
                       model=mods.geometry.hydraulic_size_factors.spheres_and_cylinders, 
                       pore_diameter='pore.diameter', 
                       throat_diameter='throat.diameter')

geo = MyGeo(network=pn, pores=pn.Ps, throats=pn.Ts, random_state = 123) 
geo.show_hist('pore.diameter')
geo.show_hist('throat.diameter')

pn['pore.top'] = pn.coords[:, 2] + geo["pore.diameter"] / 2 >= 4.40*(10**-3)
pn['pore.bottom'] = pn.coords[:, 2] - geo["pore.diameter"] / 2 <= 0.5*(10**-3)

op.topotools.plot_coordinates(network=pn, pores=pn.pores('top'), 
                              markersize=50, c='r')

op.topotools.plot_coordinates(network=pn, pores=pn.pores('bottom'), 
                              markersize=50, c='b')

water = op.phases.Water(network=pn) # Phase Object

phys_water =op.physics.Standard(network=pn, phase=water
                                      ,geometry=geo)

# Perform Stokes flow to find Permeability coefficient

sf_z = op.algorithms.StokesFlow(network=pn, phase=water) 
Pin = 101325
Pout = 0

sf_z.set_value_BC(pores=pn.pores('top'), values=Pin)
sf_z.set_value_BC(pores=pn.pores('bottom'), values=Pout)
sf_z.run()
A = 1.98*(10**-5) # In metri
L = 4450*(10**-6) # In metri
mu = water['pore.viscosity'][0] # Water Viscosity in Pa*s
Q = sf_z.rate(pores=pn.pores('top'), mode='group')                                        
Kz = (Q*L*mu/(A*(Pin - Pout))) 
print('The permeability coefficient is:', Kz)

n =(((sum(geo['pore.volume'])))/(87941400000*(10**-18)))*100
print('The porosity is:', n)

I obtain a Kz=10^(-20) m/s but, from experimental data, I kwnow it must be equal to 10^(-12) m/s. How is it possible?

**Comment by HFenice** [comment]
I also don't understand the difference between throat hydraulic conductance and throat hydraulic size factors. Can you help me?

**Reply by jgostick** [reply]
Hi @HFenice , let me answer your 2nd question first:  'hydraulic conductance' (``g``) is the value that goes into the equation: ``Q__ij = g Delta P``, where ``Q_ij`` is the volumetric flow rate between pores i and j, and ``Delta P`` is the pressure difference between pores i and j.  That is all.  How you get ``g`` is very crucial to PNM and is probably why your values are off, which I'll address below.

UPDATE:  I forget to mention that ``size_factors`` are a new concept we started using to account for the geometrical aspects of the pores and throats, like shape, converging/diverging, etc.  When computing the 'hydraulic conductance' for a throat you need both the *size factor* part (``Sf``), and the *fluid properties* part (eg viscosity ``mu``), so ``g = Sf / mu``.

**Comment by jgostick** [comment]
You have a lot of code in your question so its hard for me to understand at a quick glance (which is all I have time for at the moment)...however, from what I can tell you are generating a network using Garbriel, then testing its K value.  For this to match the experimental value you must 'calibrate' the network, meaning you need to find pore and throat sizes that match your material.  I direct you to two of my old papers where I explain this process in detail for a [cubic network](https://doi.org/10.1016/j.jpowsour.2007.04.059) and a [Delaunay network](https://doi.org/10.1149/2.009308jes), similar to the Gabriel network you are working on.  I can also point you to a tutorial we have about [adjusting the sizes](https://github.com/PMEAL/OpenPNM/blob/organize_examples/examples/tutorials/geometry/adjusting_pore_size_distributions.ipynb) of your pore network to match desired distributions.

**Reply by HFenice** [reply]
Many thanks for your support. The problem is that I have already used the pore size and the throat size of my material but the computational K value does not match with the experimental one.

**Reply by jgostick** [reply]
Where did you get the pore and throat size info? Does the Porosimetry simulation match the mercury intrusion data?

**Reply by HFenice** [reply]
My material is made of bubbles (=pores) immersed in a cemented, homogeneous and microporous matrix (that I represent with throats) that connect them. I have the information about pore size from a microtomography scan, while I know the throat size (that I set as a constant) thanks to mercury intrusion data (I set throat.diameter equal to the mean of the micropore size distribution that I got from mercury intrusion tests).

**Reply by jgostick** [reply]
Since you have a tomography scan, have you considered doing a network extraction using PoreSpy?  We have the 'snow' algorithm there.

**Reply by jgostick** [reply]
Anyway, you should run the Porosimetry algorithm and compare it to the mercury data that you have.  Also, how confident are you in the K=10^12 value? That seems quite high considering the structure you describe.

**Reply by HFenice** [reply]
I found out the mistake. It was a problem of units. The last question: is the StokesFlow alghoritm used for saturated media right? Many thanks in advance.

**Reply by jgostick** [reply]
It can be used for either.  You can account for the presence of non-flowing phase in the pores/throats by adjusting the hydraulic conductance of the filled pores.  Here is an example of using FickianDiffusion of gas through a partially water filled network:  https://github.com/PMEAL/OpenPNM/blob/dev/examples/notebooks/algorithms/multiphase/relative_diffusivity.ipynb.
